clash-lib/src/Data/Aeson/Extra.hs

```diff
@@ -4,47 +4,118 @@
-replaceCommonEscapes = ( T.replace (pack "\\n") (pack "\n") ) .
-                       ( T.replace (pack "\\\\") (pack "\\") ) .
-                       ( T.replace (pack "\\\"") (pack "\"") )
+-- | See 'toSpecNewlines'. A line map maps "virtual" lines to a range of
+-- "real" lines.
+type LineMap = IntMap (Int, Int)
+
+-- | Aeson versions <1.5.0 accept unescaped newlines in JSON strings. This is in
+-- violation of RFC 7159. Aeson 1.5.0 fixes this bug. Unfortunately, "Clash
+-- JSON" files rely on the old behavior. This function replaces newlines (in
+-- stings) with their escaped variants.
+toSpecNewlines
+  :: ByteString
+  -> Either UnicodeException (LineMap, ByteString)
+toSpecNewlines bs = do
+ s0 <- LT.unpack <$> LT.decodeUtf8' bs
+ Right ( toLineMap (go2 0 False s0)
+       , LT.encodeUtf8 (LT.pack (go False s0)))
+ where
+  -- replace newlines with escaped ones
+  go :: Bool -> String -> String
+  go _ [] = []
+  go True ('\n':rest) = '\\' : 'n' : go True rest
+  go inString ('\\':r:rest) = '\\' : r : go inString rest
+  go inString ('"':rest) = '"' : go (not inString) rest
+  go inString (r:rest) = r : go inString rest
+
+  -- Calculate real:virtual mapping
+  go2
+    -- virtual line counter.
+    :: Int
+    -- Processing a JSON string?
+    -> Bool
+    -- String left to process
+    -> String
+    -- Virtual line numbers. [0, 1, 1, 2, 2, ..] would mean:
+    --
+    --  real | virtual
+    --  --------------
+    --     0 |       0
+    --     1 |       1
+    --     2 |       1
+    --     3 |       2
+    --     4 |       2
+    --   ... |     ...
+    --
+    -> [Int]
+  go2 n _        [] = [n]
+  go2 n True     ('\n':rest) = n : go2 n True rest
+  go2 n False    ('\n':rest) = n : go2 (succ n) False rest
+  go2 n inString ('\\':_:rest) = go2 n inString rest
+  go2 n inString ('"':rest) = go2 n (not inString) rest
+  go2 n inString (_:rest) = go2 n inString rest
 
-genLineErr' :: [Text] -> (Int, Int) -> Int -> Text
-genLineErr' allLines range errorLineN = T.unlines [ T.concat [ if i == errorLineN then pack ">> " else  pack "   "
-                                                             , pack $ show i
-                                                             , pack ". "
-                                                             , allLines !! i
-                                                             ] | i <- Ix.range range]
+  toLineMap :: [Int] -> LineMap
+  toLineMap virtuals =
+      IntMap.fromList
+    $ map (second (\reals -> (minimum reals, maximum reals)))
+    $ map (\(unzip -> (virts, reals)) -> (head virts, reals))
+    $ groupBy (\a b -> fst a == fst b)
+    $ zip virtuals [(0::Int)..]
+
+genLineErr' :: [Text] -> (Int, Int) -> [Int] -> Text
+genLineErr' allLines range errorLines =
+  T.unlines [ T.concat [ if elem i errorLines then pack ">> " else  pack "   "
+            , pack $ show (i + 1)
+            , pack ". "
+            , allLines !! i
+            ] | i <- Ix.range range]
```

Comment:
I'd suspect doing `drop` and `take` on `allLines`  would be slightly faster and more "proper" FP.
But since this code isn't on any critical path it doesn't matter

Reply:
Yeah I thought about changing it, but wasn't worth the effort for the reason you mention.